prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+                per_device_eval_batch_size=trial.suggest_categorical(
+                    "per_device_eval_batch_size",
+                    hyperparameter_space["per_device_train_batch_size"],
+                ),
+            )
+            objective_trainer = transformers.Trainer(
+                model=model,
+                args=training_args,
+                data_collator=transformers.DataCollatorForSeq2Seq(
+                    tokenizer=self.generation_model_trainer.tokenizer
+                ),
+                train_dataset=train_dataset,
+                eval_dataset=validation,
+            )
+
+            _ = objective_trainer.train()
+            to_optimize = objective_trainer.evaluate()
+            return to_optimize["eval_loss"]
+
+        study = optuna.create_study(
+            study_name="automatic_hyperparameter_search", direction="minimize"
+        )
+
+        study.optimize(func=objective, n_trials=self.n_trials, gc_after_trial=True)
+        best_hyperparameters = {
+            "learning_rate": float(study.best_params["learning_rate"]),
+            "weight_decay": float(study.best_params["weight_decay"]),
+            "num_train_epochs": int(study.best_params["num_train_epochs"]),
+            "save_strategy": study.best_params["save_strategy"],
+            "per_device_train_batch_size": study.best_params[
+                "per_device_train_batch_size"
+            ],
+        }
+        return best_hyperparameters
+
+    def select_from_hyperparameters(
+        self,
+        training_datasets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: Optional[dict[str, Any]] = None,
+    ) -> tuple[PreTrainedModel, PreTrainedTokenizer]:
+        """Select a model among a set of hyperparameters (given or inferred). # noqa D410
+
+        Args:
+            training_datasets: One or more training datasets for the trainer.
+            validation: A dataset for computing validation metrics.
+            hyperparameters: A dictionary of hyperparameter choices.
+
+        If no hyperparameter_space is choosen, then the default hyperparameter_space
+        will be choosen. Here is the example of how the space looks like:
+        hyperparameter_space = {
+            "min_num_train_epochs": 5,
+            "max_num_train_epochs": 10,
+            "save_strategy": ["epoch", "steps", "no"],
+            "evaluation_strategy": ["epoch", "no"],
+            "per_device_train_batch_size": [4, 8, 16, 32],
+            "min_weight_decay": 1e-5,
+            "max_weight_decay": 1e-1,
+            "min_learning_rate": 1e-5,
+            "max_learning_rate": 1e-1,
+        }
+        Return:
+            A model and tokenizer (with hyperparameters from given range).
+        """
+        model = self.generation_model_trainer.model
+        tokenizer = self.generation_model_trainer.tokenizer
+
+        # FIXME: Needs discussion where to keep all of these paths
+        best_model_path = "./best_model"
```

Comment:
We could use `Path("result/trained_model")`, which is where the demo script currently saves the trained model

Reply:
Ohh sure, we can do that. I could have committed the changes, in case if you suggested here.